From 185b65bfc3d50b3435630e7211bf2181ba31b218 Mon Sep 17 00:00:00 2001
From: dkolhe <dkolhe@maxlinear.com>
Date: Tue, 3 Feb 2026 02:32:05 -0800
Subject: [PATCH] WLANRTSYS-99764 hostapd: Fix CCA event handling to avoid
 invalid HE BSS color
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

CCA events were handled by both hostapd_event_color_change and the
deprecated hostapd_event_cca

hostapd_event_color_change resets hapd->cca_color after beacon
update. Reusing the same hapd context in hostapd_event_cca propagated
a zero value to he_bss_color, which was later sent to the kernel and
rejected as out of valid range (1–63)

Fix: Align CCA events handling to use proprietary implemenatation of
hostapd_event_cca

Signed-off-by: dkolhe <dkolhe@maxlinear.com>
---
 src/ap/drv_callbacks.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 9469a6f7d..a778d9294 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -3246,23 +3246,31 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 					data->bss_color_collision.link_id);
 		wpa_printf(MSG_DEBUG, "CCA started on %s",
 			   hapd->conf->iface);
+#ifdef CONFIG_VENDOR_MXL
 		hostapd_event_cca(hapd, event);
+#endif /* CONFIG_VENDOR_MXL */
 		break;
 	case EVENT_CCA_ABORTED_NOTIFY:
 		hapd = switch_link_hapd(hapd,
 					data->bss_color_collision.link_id);
 		wpa_printf(MSG_DEBUG, "CCA aborted on %s",
 			   hapd->conf->iface);
-		hostapd_event_color_change(hapd, false);
+#ifdef CONFIG_VENDOR_MXL
 		hostapd_event_cca(hapd, event);
+#else
+		hostapd_event_color_change(hapd, false);
+#endif /* CONFIG_VENDOR_MXL */
 		break;
 	case EVENT_CCA_NOTIFY:
 		hapd = switch_link_hapd(hapd,
 					data->bss_color_collision.link_id);
 		wpa_printf(MSG_DEBUG, "CCA finished on %s",
 			   hapd->conf->iface);
-		hostapd_event_color_change(hapd, true);
+#ifdef CONFIG_VENDOR_MXL
 		hostapd_event_cca(hapd, event);
+#else
+		hostapd_event_color_change(hapd, true);
+#endif /* CONFIG_VENDOR_MXL */
 		break;
 #endif /* CONFIG_IEEE80211AX */
 #ifdef CONFIG_IEEE80211BE
-- 
2.43.0

